### PR TITLE
Index finalized data by slot instead of block root

### PR DESCRIPTION
**Pull request by ajsutton**
## PR Description
So that we can efficiently find the closest available snapshot prior to the requested one (and then regenerate from there), switch to storing finalised states and blocks indexed by slot instead of blockRoot and then have a blockRoot->slot index instead of slot->blockRoot.

Only changes the database format for v4 which still isn't the default. Had to change how V3 looks up data but what it stores on disk is unchanged so still backwards compatible.

## Fixed Issue(s)
Part of #2198 

## Documentation

- [x] I thought about documentation and added the `documentation` label to this PR if updates are required.

### CODE REVIEW

File: storage/src/test/java/tech/pegasys/teku/storage/server/AbstractDatabaseTest.java [hunk 1]
```diff
@@ -657,13 +660,24 @@ protected void testShouldRecordFinalizedBlocksAndStates(
         break;
       case PRUNE:
         // Check pruned states
-        final List<Bytes32> unavailableRoots =
-            allBlocksAndStates.stream().map(SignedBlockAndState::getRoot).collect(toList());
+        final List<UnsignedLong> unavailableRoots =
```
Comment (mbaxter) [inline storage/src/test/java/tech/pegasys/teku/storage/server/AbstractDatabaseTest.java]:
```suggestion
        final List<UnsignedLong> unavailableSlots =
```

File: storage/src/test/java/tech/pegasys/teku/storage/server/AbstractDatabaseTest.java [hunk 2]
```diff
@@ -645,6 +647,7 @@ protected void testShouldRecordFinalizedBlocksAndStates(
     assertBlocksFinalized(expectedFinalizedBlocks);
     assertGetLatestFinalizedRootAtSlotReturnsFinalizedBlocks(expectedFinalizedBlocks);
     assertBlocksAvailableByRoot(expectedFinalizedBlocks);
+    assertFinalizedBlocksAvailableViaStream(primaryChain);
```
Comment (mbaxter) [inline storage/src/test/java/tech/pegasys/teku/storage/server/AbstractDatabaseTest.java]: This helper method is oddly specific - why not pass in `expectedFinalizedBlocks`?

File: storage/src/test/java/tech/pegasys/teku/storage/server/AbstractDatabaseTest.java [hunk 3]
```diff
@@ -657,13 +660,24 @@ protected void testShouldRecordFinalizedBlocksAndStates(
...
   }
 
+  private void assertFinalizedBlocksAvailableViaStream(final ChainBuilder primaryChain) {
+    try (final Stream<SignedBeaconBlock> stream =
+        database.streamFinalizedBlocks(UnsignedLong.valueOf(1), UnsignedLong.valueOf(3))) {
+      assertThat(stream)
+          .containsExactly(
+              primaryChain.getBlockAtSlot(1),
+              primaryChain.getBlockAtSlot(2),
+              primaryChain.getBlockAtSlot(3));
```
Comment (mbaxter) [inline storage/src/test/java/tech/pegasys/teku/storage/server/AbstractDatabaseTest.java]: Shouldn't block `7` also be available?  It was finalized.

**Review by mbaxter — APPROVED** [review]
LGTM - just left a couple of questions about tests

File: storage/src/test/java/tech/pegasys/teku/storage/server/AbstractDatabaseTest.java (hunk 3 above)
Comment (ajsutton) [reply storage/src/test/java/tech/pegasys/teku/storage/server/AbstractDatabaseTest.java]: We're specifically requesting blocks 1-3 to test the range is respected.  I've made it a more generic function which hopefully make that clearer as well.

File: storage/src/test/java/tech/pegasys/teku/storage/server/AbstractDatabaseTest.java (hunk 1 above)
Comment (ajsutton) [reply storage/src/test/java/tech/pegasys/teku/storage/server/AbstractDatabaseTest.java]: Done.

**Pull request merged**